src/gt4py/next/iterator/transforms/fuse_as_fieldop.py

```diff
@@ -204,48 +305,141 @@ def apply(
         offset_provider_type: common.OffsetProviderType,
         uids: Optional[eve_utils.UIDGenerator] = None,
         allow_undeclared_symbols=False,
+        within_set_at_expr: Optional[bool] = None,
     ):
         node = type_inference.infer(
             node,
             offset_provider_type=offset_provider_type,
             allow_undeclared_symbols=allow_undeclared_symbols,
         )
 
+        if within_set_at_expr is None:
+            within_set_at_expr = not isinstance(node, itir.Program)
+
         if not uids:
             uids = eve_utils.UIDGenerator()
 
-        return cls(uids=uids).visit(node)
+        return cls(uids=uids).visit(node, within_set_at_expr=within_set_at_expr)
+
+    def visit(self, node, **kwargs):
+        new_node = super().visit(node, **kwargs)
+        if isinstance(node, itir.Expr) and hasattr(node.annex, "domain"):
+            new_node.annex.domain = node.annex.domain
+        return new_node
+
+    def visit_SetAt(self, node: itir.SetAt, **kwargs):
+        return itir.SetAt(
+            expr=self.visit(node.expr, **kwargs | {"within_set_at_expr": True}),
+            domain=node.domain,
+            target=node.target,
+        )
+
+    def visit_FunCall(self, node: itir.FunCall, **kwargs):
```

Comment:
Let's rewrite the whole thing together for 2 reasons:
- knowledge sharing
- after it is now (maybe) correct, focus on readability.